### ISSUE TITLE
Why are LSTMs better than Elman RNNs?  Oscar Chang

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2775]
Why are LSTMs better than Elman RNNs? · Oscar Chang
    
  <br>
<br>
https://ift.tt/2T0ZtCn<br>
<br>
<br>
<br>
<div><article><h1>Why are LSTMs better than Elman RNNs?</h1>
<time datetime="2018-05-28T00:00:00+00:00">28 May 2018</time><p>
<strong>tldr;</strong> Elman RNNs suffer from the vanishing/exploding gradients problem, which hurts their ability to learn long range dependencies. The forget gate in LSTMs is the main reason why LSTMs are able to delay this problem and hence, process longer input sequences.
</p>
<p>Recurrent neural networks (RNNs) are useful for modelling sequential data of variable length, for example the words in a speech or stock price movements over a period of time. In this post, we assume that the reader is familiar with two of the most popular RNNs, the Elman RNN and the Long Short-Term Memory network (LSTM).</p>
<p>RNNs get more difficult to train as the length of the input sequence increases, which is why it is common to use truncated backpropagation through time to train RNNs.</p>
<p>The aim of this post is to explain why LSTMs can handle longer sequences than Elman RNNs.</p>
<h2>General RNN</h2>
<p>RNNs take the general form . The hidden state at each time step is a function of the hidden state and the input at the previous time step.</p>
<p>By differentiating the expression<sup></sup>, we realize that the gradient of the loss at is the product of the gradient of the loss at and the partial derivative of with respect to .</p>
<p>If on average, we end up with the exploding gradients problem, since repeated multiplication by a number above tends to infinity.</p>
<p>If on average, we end up with the vanishing gradients problem, since repeated multiplication by a positive number smaller than tends to zero.</p>
<h2>Elman RNNs</h2>
<p>Elman RNNs instantiate with the following equation.</p>
<p>Applying the chain rule, we obtain where maps a vector into its diagonal matrix form.</p>
<p><strong>The value of depends on both the weight matrix and the choice of activation function .</strong></p>
<p>Where is the identity, the gradient explodes whenever the largest eigenvalue of is above and vanishes whenever it is below .</p>
<p>This makes it difficult to learn long range dependencies, since neither infinite nor zero gradients are helpful.</p>
<h2>LSTMs</h2>
<p>LSTMs instantiate with the following set of equations.</p>
<p><em>Note: we are overloading with both the forget gate and the recurrent function . Here, the hidden (cell) state is and not , which is the output state.</em></p>
<p>We can write as a sum of the following components using the chain rule again.</p>
<p>The latter three terms in the sum depend on partial derivatives with respect to . Upon expanding those terms, we find that they are a (long) chain of weight matrices and derivatives of activation functions, as in the Elman RNN case.</p>
<p> is usually the sigmoid function, while and are usually the function. The sigmoid function has a derivative upper bounded by while the derivative of the function is at most . This makes it likely for these terms to be small.</p>
<p>But crucially, the first term is just the activation of the forget gate at that timestep. <strong>This value is mostly independent of the eigenstructure of a weight matrix and the derivative of an activation function.</strong></p>
<p>We can therefore mostly delay the vanishing/exploding gradients problem whenever stays close to even when the other three terms vanish (but not explode), which can be achieved with careful weight initialization.<sup></sup></p>
<p>Hence, the forget gate is the main reason why LSTMs are better than Elman RNNs at processing long input sequences.</p>
<p><small><em>Thanks to Jenny Chen, Kui Tang and Jaan Altosaar for reading a draft of this post</em></small></p>
</article></div><br>
<br>
<br>
<br>
<br>
<br>
via 
    
      Oscar Chang · 阳光宅男 B-log
    
   <br>
<br>
February 26, 2019 at 11:45AM
